### ISSUE TITLE
[Proposal] Enrich Pub/Sub API and feature set

### ISSUE THREAD

**Issue by artursouza** [dapr/dapr#843]
# Proposal

## Overview

The current PubSub API is extremely simple but lacks very important configurations and features. This proposal includes a set of changes to be implemented for the PubSub API.

### 1. *Missing retriable vs non-retriable error code handling*
Publishing a message to Dapr can fail but the error might not be retriable. The publishing API only supports two status codes: 200 and 500. There are many error codes that can indicate if an error is not retriable. This is very important, so the client can know that a given error is deterministic and no further retries should be done. For example, a 400 status code should indicate to the client that the request sent to Dapr is not valid. The same happens for the subscribers, when Dapr expects only 200 and 500 as valid status code from client's API.

Expeted behavior: expand the list of expected error codes and specify which errors are retriable vs non-retriable.

#### Update

Status: DONE in https://github.com/dapr/dapr/pull/1993

Retry logic is not exactly how described in this issue. Dapr's current logic to handle error codes is described here https://github.com/dapr/docs/blob/master/reference/api/pubsub_api.md

### 2. *Hard-coding of `/<topic>` for subscriber APIs*
When an applications subscribes to one or more topics, Dapr assumes one /<topic> api per topic. This is a strong assumption and it limits the design choices for the customer APIs. The client, might want to have a RESTful compliant API but Dapr would violate that with this enforcement.

Expected behavior: the subscriber response should returns a list of objects (and not a list of strings), so each subscribed topic can have many addition properties: URL endpoint, Retry Timeout, etc.

Example of subscriber's response:
```json
{
  "subscriptions": [
    {
       "topic": "hotels",
       "callback": "/dapr/orders/hotels"
    },
    { 
       "topic": "comments",
       "callback": "/dapr/reviews/comments",
       "retryTimeInMinutes": 60,
       "retryLogic": "exponentialBackoff"
    }
  ]
}
```

#### Update

Status: DONE in https://github.com/dapr/dapr/pull/1529

Custom route is supported and documented in https://github.com/dapr/docs/blob/master/reference/api/pubsub_api.md. Configuration per topic subscription can be further extended but specific new configs (like retryLogic) is not an ask here, just an example of future possibilities with this new schema.

### 3. *Client-provided message Id*
ServiceBus supports client-provided Message Id. The idea is that the publisher can take ownership of Id generation for the messages and take advantage of extra features in the underlying service. In ServiceBus, Message Id can be used to dedup messages during a time interval, providing free one-publish guarantee. Not every implementation will support client-provided Message Ids, so it would be ignored in those cases.

Expedted behavior: publisher can (optionally) provide a special attribute in the payload (e.g. "id" or "dapr.messageId") and Dapr can extract it and use it as the message id (if the component supports it). The message format can also enforce the CloudEvents spec of uniqueness of "source" + "id".

Example of publisher's message:
```json
{
  "id": 123456,
  "data": {
     "order": 42
  }
}
```

#### Update

Status: DONE in https://github.com/dapr/dapr/issues/1372

Users can accomplish this by publishing a cloud event directly to Dapr. This way, Dapr will not wrap it inside another CloudEvent, keeping the id provided by the user. Publishing the following event, for example, keeps the `id` on the subscriber side.

```json
    {
        "id" : "100",
        "specversion" : "1.0",
        "type" : "com.github.pull.create",
        "source" : "https://github.com/cloudevents/spec/pull",
        "subject" : "123",
        "time" : "2018-04-05T17:31:00Z",
        "comexampleextension1" : "value",
        "comexampleothervalue" : 5,
        "datacontenttype" : "text/xml",
        "data" : "<much wow=\"xml\"/>"
    }
```

### 4. *Message Expiration*
ServiceBus offers native support for TTL per message. Kafka does it at a global or topic level but not at message level. Dapr can use native support for some implementation but also implement TTL by not publishing messages that have expired. The same will happen if a message cannot be sent to a subscriber after retrying for a configurable amount of time.

Expected behavior: Publisher can (optionally) enter the expiration date/time in UTC as part of a CloudEvent payload (e.g. "daprMessageExpiration") and Dapr will not publish (or retry) the message after the given expiration.

Example of publisher's message:
```json
    {
        "id" : "100",
        "specversion" : "1.0",
        "type" : "com.github.pull.create",
        "source" : "https://github.com/cloudevents/spec/pull",
        "subject" : "123",
        "time" : "2018-04-05T17:31:00Z",
        "daprMessageExpiration" : "2020-04-05T17:31:00Z",
        "comexampleothervalue" : 5,
        "datacontenttype" : "text/xml",
        "data" : "<much wow=\"xml\"/>"
    }
```

#### Update

Not done yet. Tracking in #2216 

### 5. *DeadLetter Topic*
After a message cannot be delivered to a subscriber (retry time out or message expiration), the message should be sent to a special topic. The subscriber can subscribe from that topic as well, or simply leave messages there to be manually verified. Dapr will try to publish to dead letter topic only once. This is a way the subscriber can recover if the API was returning non-retriable errors wrongly due to a code bug, for example, or messages expired after being retried for too long. This allows subscribers to self-recover without escalating to the publisher.

Expected behavior: (1) Subscriber can configure a deadletter topic for a given subscription. (2) Subscribers can subscribe to a deadletter topic like any other topic.

Example of subscriber's response:
```json
{
  "subscriptions": [
    {
       "topic": "hotels",
       "callback": "/dapr/orders/hotels"
    },
    { 
       "topic": "comments",
       "callback": "/dapr/reviews/comments",
       "deadLetterTopic": "comments_myconsumergroup_deadletter",
    },
    { 
       "topic": "comments_myconsumergroup_deadletter",
       "callback": "/dapr/reviews/comments_deadletter",
    }
  ]
}
```

#### Update

Not done yet. Tracking in #2217

### 6. *Queue support*
The current PubSub model is a broadcast model because one published message will be consumed (ideally) by all subscribers - except when there is a major outage by one of them. SQS, for example, is a queue - it means at least one of the subscribers will receive that message but not necessarily more. Dapr should enable both PubSub and Queue patterns. A Queue should not allow more than one subscriber application. An application can have multiple nodes, so they take care of redundancy under a given application.

Expected behavior:  (1) Create a Queue component in Dapr.

#### Update

Status: ARCHIVED as per https://github.com/dapr/dapr/issues/771

This feature can be implemented on each component. PubSub can behave as a topic or queue depending on the implementation. In EventHub, for example, the consumerGroupId can define the behavior of queue vs topic.

### 7. *Batching*

The current PubSub model notifies consumers by making a roundtrip per message. A way to improve message throughput would be to allow consumers to declare the intention of consuming messages in batches. Processing messages in batches lead to the following optimizations:
- roundtrip reduction between dapr and consuming component by sending messages batches
- consumer computation by observing a message batch instead of a single item

Expected behavior: Subscribers can (optionally) choose to receive messages in batches. 

Example of a subscriber's response where batching is requested:

```
{
  "subscriptions": [
    {
       "topic": "A",
       "maxBatchSize": "10"
    }
  ]
}
```

The optional parameter `maxBatchSize` defines the maximum number of messages to be delivered in a single receive loop. By default, maxBatchSize should be 1.

Message checkpointing in batch mode are for all messages (all succeeded or failed).

Distributed tracing should track each message inside the batch in such a way that users can still find them in the target telemetry backend. In other words, if messages A, B and C were processed in the same batch, searching for message B telemetry should return the batch execution.

In order to enable receiving multiple messages the subscriber callback contract could be changed from:

```go
// NewMessage is an event arriving from a message bus instance
type NewMessage struct {
	Data     []byte            `json:"data"`
	Topic    string            `json:"topic"`
	Metadata map[string]string `json:"metadata"`
}
```

to

```go
// Message is one event arriving from a message bus instance
type Message struct {
	Data []byte `json:"data"`
	Metadata map[string]string `json:"metadata"`
}

// NewMessages contains one or more events arriving from a message bus instance for a given topic
type NewMessages struct {
	Messages []Message `json:"messages"`
	Topic    string    `json:"topic"`
}
```

#### Update

Not done yet. Tracking in #2218

### 1. Should each component implementation handle dead letter queue instead?

Dapr does not stop implementations from handling deadletetter queue or topic, but dapr can offer this feature even to components that do not this feature built in. This deadletter topic goes side by side with message expiration feature.

### 2. Should Dapr delegate batching to component implementation?

Without this feature, components can still support batching via metadata *but* the messages will still be sent to subscribers one-by-one. Having this feature built in Dapr (breaking the API) will allow apps to receive messages in actual batches (one request with many events). 

## Archived

This issue is being closed and archived since we now have individual items tracking each feature.

**Comment by artursouza** [comment]
@yaron2

**Comment by dev-nextprogress** [comment]
+1
Great proposals Artur

**Comment by yaron2** [comment]
Thanks for this!

*Edit Note: I edited the title to reflect better the contents, which contains additions of functionality and not modifications or reductions of the existing APIs*

As each of the items here will contain at least a single PR, you or anyone else picking work on any item should create a dedicated issue, and we could use this issue for tracking or reference.

> 1. Missing retriable vs non-retriable error code handling

This is a low hanging fruit and I think the best item to start with. If/when you open a new issue for it, please mark it with a `help-wanted` or `good-first-issue` label.

> 2. Hard-coding of /<topic> for subscriber APIs

lgtm, but I'd remove (for now) the fields for retries from the example as they do not have direct correlation to the proposal here. In the spirit of small focused deliverables, I suggest to focus this proposal on the user code subscription API and in the future add more functionality to it separately.

> 3. Client-provided message Id

For me, this is a nice to have feature that is also implementation dependent, and thus classified as best effort.
I would encourage to think whether Dapr needs to implement it's own dedupe mechanism to avoid the best-effort case. Again, if/when this becomes a thing, should be discussed in a separate issue.

> 4. TimeToLive and RetryTimeOut

Same as the above. If we can implement TTL ourselves that's the preferred route. Not to go into implementation details here, the TTLs would need to be stored somewhere, and we should be careful to try and not create coupling between state and pub/sub. 

> 5. DeadLetter Topic

This is a VERY important feature we need to have. Luckily, Dapr today can build on it's own APIs to avoid from making this a best effort or implementation specific.

> 6. Queue support

You didn't take into consideration the existence of consumer groups and consumer IDs in Dapr's pub/sub.
Dapr Pub/Sub ideally (the case for Redis Streams and Azure Service Bus) turns a Dapr ID into a consumer ID, making sure that the correct behavior is that multiple consumers within a consumer group will get the message once, and that each unique consumer group will pull the message once.

The case for the current NATS implementation (not to be confused with NATS streaming, not yet implemented) and the upcoming RabbitMQ PR is that of queues, meaning consumerIDs are not taken into account, making the /publish method act more like /queue and /subscribe to /dequeue.

In this case, two Dapr apps with two IDs listening on the same topic/queue will only see one of them receive the message instead of both, as as in the Redis Streams and Azure Service Bus implementations.

So, this presents us with the following options:

* We introduce a new Queue building block API and separate the queue implementations (RabbitMQ, NATS, etc.) from Pub/Sub

* We keep the implementations under Pub/Sub, but make it explicit in the docs about which implementations work as streaming pipelines and which work as queues

**Comment by fbeltrao** [comment]
What about including batching in this discussing? 
Batching would be an option for scenarios where a high data volume needs to be processed (binding or pub/sub), aiming to reduce the amount of roundtrips required to process them (by sending messages in batch to consumers).

The challenges I can see are:
- Message ack: either all or nothing
- Distributed tracing: how to built trace if we handling messages in bulk?

**Comment by artursouza** [comment]
I agree that we should handle batching too. @fbeltrao , can you edit the proposal to include batching and the expected behavior in Dapr?

**Comment by fbeltrao** [comment]
_**7. Batching**_

The current PubSub model notifies consumers by making a roundtrip per message. A way to improve message throughput would be to allow consumers to declare the intention of consuming messages in batches. Processing messages in batches lead to the following optimizations:
- roundtrip reduction between dapr and consuming component by sending messages batches
- consumer computation by observing a message batch instead of a single item

Expected behavior: Subscribers can (optionally) choose to receive messages in batches. 

Example of a subscriber's response where batching is requested:

```
{
  "subscriptions": [
    {
       "topic": "A",
       "maxBatchSize": "64"
    }
  ]
}
```

The optional parameter `maxBatchSize` defines the maximum message amount to be delivered in a single receive loop. The absence of a value for `maxBatchSize` indicates that messages will be sent individually (current behavior).

Message check pointing in batch mode are for all messages (all succeeded or failed).

Distributed tracing should track each message inside the batch in such a way that users can still find them in the target telemetry backend. In other words, if messages A, B and C were processed in the same batch, searching for message B telemetry should return the batch execution.

In order to enable receiving multiple messages the subscriber callback contract could be changed from:

```go
// NewMessage is an event arriving from a message bus instance
type NewMessage struct {
	Data  []byte `json:"data"`
	Topic string `json:"topic"`
}
```

to

```go
// Message is one event arriving from a message bus instance
type Message struct {
	Data []byte `json:"data"`
}

// NewMessages contains one or more events arriving from a message bus instance for a given topic
type NewMessages struct {
	Messages []Message `json:"messages"`
	Topic    string    `json:"topic"`
}
```

**Comment by artursouza** [comment]
@fbeltrao  Added batching to the proposal.

Also, your partition feature seems to be similar (not the same) as item number 3 here. It is a way for Dapr to understand special attributes in the published message and handle that differently. Some of them will be specific to the component implementation (Kafka vs ServiceBus have different feature sets).

**Comment by fbeltrao** [comment]
Thanks @artursouza. We could include in _3. Client-provided message Id_ the option to define the partition key (for EventHubs, Kafka and others that support partitioning). By defining a partition key a producer can ensure event ordering for related items (i.e. events from user '123' would end up in same partition).

Options:

a) Add to publisher message contract, allowing strings or numbers
```json
{
  "partitionKey": 123456,
  "data": {
     "order": 42
  }
}
```

b) Add optional "partitionKey" in output binding/publisher metadata to define a value. This would handle scenarios where the key is a string. Scenarios where the partition key is another type wouldn't be supported.

**Comment by artursouza** [comment]
BTW, Maybe you can tackle this as part of item number 3. It will create a way for the publisher to add optional special fields that Dapr will understand. One of them is PartitionKey, another one is MessageID and more to come. I am all in for that, @yaron2  also mentioned that the error handling is also another low hanging fruit.

**Comment by yaron2** [comment]
Crossing beams with https://github.com/dapr/dapr/issues/1042#issuecomment-586978979

**Comment by denniszielke** [comment]
What is your thinking on supporting topic filters (https://docs.microsoft.com/en-us/azure/service-bus-messaging/topic-filters)? 
We are building a scenario which will depend on broadcasted messages that are very closely related -so it does not make much sense to distribute them across different topics. But we want certain subscribers be able to ignore parts of them to reduce the amount of executions. So a filter would help us a lot.

**Comment by yaron2** [comment]
> What is your thinking on supporting topic filters (https://docs.microsoft.com/en-us/azure/service-bus-messaging/topic-filters)?
> We are building a scenario which will depend on broadcasted messages that are very closely related -so it does not make much sense to distribute them across different topics. But we want certain subscribers be able to ignore parts of them to reduce the amount of executions. So a filter would help us a lot.

@denniszielke  That makes sense.

According to the link above, ASB cannot filter according to the message body.
Dapr can actually do this, provided the data is JSON.

Which properties were you thinking of filtering by that would be useful?

**Comment by denniszielke** [comment]
We were thinking of using user defined application properties that we would use in the correlation filter. That would also require to make it possible to attach user defined application properties during the publish of messages via dapr.

**Comment by yaron2** [comment]
Do you have an idea of how you'd like the see the developer tell Dapr about the properties to filter by? Is this metadata property on the component YAML, or is it a new CRD?

**Comment by msfussell** [comment]
@artursouza can you revisit this and determine what has been covered and what remains?

**Comment by XavierGeerinck** [comment]
+1 on the queue requirement! :) would be able to utilize this!

**Comment by artursouza** [comment]
> +1 on the queue requirement! :) would be able to utilize this!

Which component are you using? This might be already available if the component has support for consumerGroup ids.

**Comment by msfussell** [comment]
@artursouza - Can we close this issue now separate ones have been created?

**Comment by XavierGeerinck** [comment]
> > +1 on the queue requirement! :) would be able to utilize this!
> 
> Which component are you using? This might be already available if the component has support for consumerGroup ids.

Sorry for the late response, I've been in Turkey at the moment with low access to internet for family visit. 

The solution I am now using is by switching it to queues (where it is supposed to be actually) and that did the trick in my use case. I have a blog post coming out soon :)

**Comment by artursouza** [comment]
@msfussell  I agree. In this case, each feature is independent and is already being tracked separately.
